### PR TITLE
Handling HTTP redirect from the docker registry v2

### LINKED ISSUE #641: dockerClient can't handle registry which is requires authorization and using HTTP redirection.
Registry could use HTTP redirects by 3xx status code for example for load balancing. If such registry requires authorization at the same time dockerClient is fails to work with it. It's successfully get auth token but fails to deliver it to the endpoint.
Because of golang http Client which is blocking "Authorization" header when following redirects.
https://golang.org/pkg/net/http/#Client

### PR DESCRIPTION
**Pull request by nikitych**
resolve containers/container-libs#271

### CODE REVIEW

**Comment by giuseppe** [comment]
the patch looks reasonable to me.

@vrothberg @mtrmac @rhatdan can we merge it?

**Review by vrothberg — COMMENTED** [review]
LGTM but I'd like @mtrmac to weigh in

**Comment by mtrmac** [comment]
This looks unlikely to be correct me:
- No such behavior is documented in https://github.com/docker/distribution/blob/master/docs/spec/api.md
- I can’t find any equivalent behavior in docker/docker
- If the server is redirecting responses for individual requests from user-specified-URL to some other URL, is there a good reason to expect that
  1. The redirect of a ping to `/v2/` would at all be an URL ending with `/v2/`?
  2. All of the future requests would be redirected to the same host? Especially considering the spec explicitly suggests redirects for pulling layers, and it would make sense to distribute the layer storage across several servers.

If I am reading docker/distribution/registry/client/auth correctly (needs checking in practice), that code instead modifies the transport to add the credentials to every request, including redirects.

**Comment by nikitych** [comment]
[](url)
I agree that API itself not take into account redirections but it's using http transport which has them in [protocol](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections).

Docker itself have a little more complex [mechanisms to handle Authorization](https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/auth/session.go#L68-L108).
They inject an `Authorization` header only for requests that match to authorized ping endpoint. Although this code will work only when `/v2/` in a path.

Now, how they use it. The [Authorizer is created at](https://github.com/moby/moby/blob/master/registry/auth.go#L192) with [challengeManager from pinger](https://github.com/moby/moby/blob/master/registry/auth.go#L175). To get credentials that pinger use [response from ping endpoint](https://github.com/moby/moby/blob/master/registry/auth.go#L288) which has already gone through all the redirects and point to the final location.

Even though [challengeManager is map](https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/auth/challenge/authchallenge.go#L52-L87) and therefore can handle multiple endpoints [AddResponse](https://github.com/moby/moby/search?l=Go&q=AddResponse) called only for ping one.

I played with docker-ce 18.09.7 and it's pass auth only when get redirected to same endpoint as ping was.
<details><summary>Request log</summary><p>

```
==> GET /v2/ HTTP/1.1
> Host: localhost4:3333
> User-Agent: docker/18.09.7 go/go1.10.8 git-commit/2d0083d kernel/4.14.35-1902.3.1.el7uek.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/18.09.7 \(linux\))
> Accept-Encoding: gzip
> Connection: close
"GET /v2/ HTTP/1.1" 307 -
< b'HTTP/1.0 307 Temporary Redirect\r\n'
< b'Server: BaseHTTP/0.6 Python/3.6.3\r\n'
< b'Date: Tue, 16 Jul 2019 18:25:39 GMT\r\n'
< b'Location: http://localhost:3333/v2/\r\n'
<== DONE
==> GET /v2/ HTTP/1.1
> Host: localhost:3333
> User-Agent: docker/18.09.7 go/go1.10.8 git-commit/2d0083d kernel/4.14.35-1902.3.1.el7uek.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/18.09.7 \(linux\))
> Referer: http://localhost4:3333/v2/
> Accept-Encoding: gzip
> Connection: close
"GET /v2/ HTTP/1.1" 401 -
< 401 http://localhost:5000/v2/
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Www-Authenticate: Basic realm="Registry Realm"
< X-Content-Type-Options: nosniff
< Date: Tue, 16 Jul 2019 18:25:39 GMT
< Content-Length: 87
< Connection: close
<== DONE
==> GET /v2/alpine/manifests/latest HTTP/1.1
> Host: localhost4:3333
> User-Agent: docker/18.09.7 go/go1.10.8 git-commit/2d0083d kernel/4.14.35-1902.3.1.el7uek.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/18.09.7 \(linux\))
> Accept: application/json
> Accept: application/vnd.oci.image.manifest.v1+json
> Accept: application/vnd.oci.image.index.v1+json
> Accept: application/vnd.docker.distribution.manifest.v2+json
> Accept: application/vnd.docker.distribution.manifest.list.v2+json
> Accept: application/vnd.docker.distribution.manifest.v1+prettyjws
> Accept-Encoding: gzip
> Connection: close
"GET /v2/alpine/manifests/latest HTTP/1.1" 307 -
< b'HTTP/1.0 307 Temporary Redirect\r\n'
< b'Server: BaseHTTP/0.6 Python/3.6.3\r\n'
< b'Date: Tue, 16 Jul 2019 18:25:39 GMT\r\n'
< b'Location: http://localhost:3333/v2/alpine/manifests/latest\r\n'
<== DONE
==> GET /v2/alpine/manifests/latest HTTP/1.1
> Host: localhost:3333
> User-Agent: docker/18.09.7 go/go1.10.8 git-commit/2d0083d kernel/4.14.35-1902.3.1.el7uek.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/18.09.7 \(linux\))
> Accept: application/json
> Accept: application/vnd.oci.image.manifest.v1+json
> Accept: application/vnd.oci.image.index.v1+json
> Accept: application/vnd.docker.distribution.manifest.v2+json
> Accept: application/vnd.docker.distribution.manifest.list.v2+json
> Accept: application/vnd.docker.distribution.manifest.v1+prettyjws
> Authorization: Basic dGVzdHVzZXI6UGFzc3cwcmQh
> Referer: http://localhost4:3333/v2/alpine/manifests/latest
> Accept-Encoding: gzip
> Connection: close
"GET /v2/alpine/manifests/latest HTTP/1.1" 200 -
< 200 http://localhost:5000/v2/alpine/manifests/latest
< Content-Length: 7424
< Content-Type: application/vnd.docker.distribution.manifest.v1+prettyjws
< Docker-Content-Digest: sha256:62e1adf948847a9578bb33a086a9630966c048a8268392500a5144a70df62a6f
< Docker-Distribution-Api-Version: registry/2.0
< Etag: "sha256:62e1adf948847a9578bb33a086a9630966c048a8268392500a5144a70df62a6f"
< X-Content-Type-Options: nosniff
< Date: Tue, 16 Jul 2019 18:25:39 GMT
< Connection: close
<== DONE
==> GET /v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde HTTP/1.1
> Host: localhost4:3333
> User-Agent: docker/18.09.7 go/go1.10.8 git-commit/2d0083d kernel/4.14.35-1902.3.1.el7uek.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/18.09.7 \(linux\))
> Accept-Encoding: identity
> Connection: close
"GET /v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde HTTP/1.1" 307 -
< b'HTTP/1.0 307 Temporary Redirect\r\n'
< b'Server: BaseHTTP/0.6 Python/3.6.3\r\n'
< b'Date: Tue, 16 Jul 2019 18:25:39 GMT\r\n'
< b'Location: http://localhost.localdomain:3333/v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde\r\n'
<== DONE
==> GET /v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde HTTP/1.1
> Host: localhost.localdomain:3333
> User-Agent: docker/18.09.7 go/go1.10.8 git-commit/2d0083d kernel/4.14.35-1902.3.1.el7uek.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/18.09.7 \(linux\))
> Accept-Encoding: identity
> Referer: http://localhost4:3333/v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde
> Connection: close
"GET /v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde HTTP/1.1" 401 -
< 401 http://localhost:5000/v2/alpine/blobs/sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Www-Authenticate: Basic realm="Registry Realm"
< X-Content-Type-Options: nosniff
< Date: Tue, 16 Jul 2019 18:25:39 GMT
< Content-Length: 149
< Connection: close
<== DONE
```
</p></details>

**Comment by mtrmac** [comment]
Yes, that’s the code I was pointing at, and your logs show the behavior: the original host, named in the user-specified reference, is in individual control of where to redirect every single request, if it should be redirected at all. I think we’ll need to achieve similar behavior (though we can probably do without most of those abstractions and indirections).

**Comment by rhatdan** [comment]
@nikitych are you still working on this one?

**Comment by nikitych** [comment]
@rhatdan I'm not sure what work I am expected to do? As far as I understand @mtrmac rejected the current patch and he will address #641 
> without most of those abstractions and indirections

Or I get it wrong?

**Comment by mtrmac** [comment]
I’m not currently working on #641 ; contributions would certainly be welcome.

The
> without most of those abstractions and indirections

comment is regarding the structure of the moby/moby+docker/docker code, with the `Authorizer`+`challengeManager` and all of that, vs the ~single-file `docker_client.go` in this project.

**Comment by nikitych** [comment]
In that case could you please describe a little more detailed what's wrong with proposed solution?
And how do you see the architecture fixing the #641?

**Comment by mtrmac** [comment]
https://github.com/containers/image/pull/645#issuecomment-511514169 lists the unwarranted assumptions of this approach, and that + https://github.com/containers/image/pull/645#issuecomment-511932969 point at the Docker authentication behavior, which I think we should broadly replicate (i.e. implement a `http.RoundTripper` that injects the credentials to all requests, including the destinations of redirects.)

**Comment by nikitych** [comment]
> that injects the credentials to all requests, including the destinations of redirects.

But it's NOT injects the credentials to all requests.
Docker add `Authorization` header only to requests whose addresses start match the endpoints with challenges. [1](https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/auth/session.go#L89-L105)
And there is only one endpoint with challenges, [the one was got by ping request](https://github.com/moby/moby/blob/master/registry/auth.go#L288)

In other words, docker adds secrets only to requests falling into the same target that was achieved during auth to ping URL. This is equivalent to if we took registry ping redirected location as truly registry. Yes, in this case we don't follow redirection on every request like docker does, but if the registry tries to redirect to location other than ping endpoint, it also won't be able to provide auth.

**Comment by mtrmac** [comment]
> > that injects the credentials to all requests, including the destinations of redirects.
> 
> But it's NOT injects the credentials to all requests.
> Docker add `Authorization` header only to requests whose addresses start match the endpoints with challenges. [1](https://github.com/moby/moby/blob/master/vendor/github.com/docker/distribution/registry/client/auth/session.go#L89-L105)
> And there is only one endpoint with challenges, [the one was got by ping request](https://github.com/moby/moby/blob/master/registry/auth.go#L288)

Thanks, that’s an important point.

**Comment by hobbytp** [comment]
What is the latest status, will this solution be merged or not?

**Comment by jankaluza** [comment]
Hi, and thank you for your contribution!

We’ve recently migrated this repository into a new **monorepo**: [containers/container-libs](https://github.com/containers/container-libs) along with other repositories

As part of this migration, this repository is no **longer accepting new Pull-Requests** and therefore this Pull-Request is being closed.

Thank you very much for your contribution. We would appreciate your continued help in migrating this PR to the new [container-libs repository](https://github.com/containers/container-libs). Please let us know if you are facing any issues.

You can read more about the migration and the reasoning behind it in our blog post: [Upcoming migration of three containers repositories to monorepo](https://blog.podman.io/2025/08/upcoming-migration-of-three-containers-repositories-to-monorepo/).

Thanks again for your work and for supporting the containers ecosystem!
